### PR TITLE
Don't show Invalid questions for Joins in Query Builder

### LINKED ISSUE #44974: Entity picker surfaces questions that are not valid joins
### Describe the bug

_Arakaki comment: this is not only about recents, the collection navigation also shows questions and models in other databases ([context](https://metaboat.slack.com/archives/C0645JP1W81/p1720027645702959))_ 

When joining data in a table, the entity picker surfaces recent questions and models from different databases. 

Until we have https://github.com/metabase/metabase/issues/3953, that is always going to lead to errors, [so as @mazameli mentioned](https://github.com/metabase/metabase/issues/38989#issuecomment-2070982901), it would be better to now show these suggestions in the entity picker at all when joining.

### To Reproduce

1. Set up two databases (locally the sample database and Postgres will work for example)
2. Create a new question in the Postgres database (details do not matter) and save it as "Question A"
3. Create a new question that has a table from the Sample database as source
4. Click Join
5. The entity picker shows "Question A" under the "Recents" tab

For this issue to come up it seems like both databases need to have questions in the recents tab.

### Expected behavior

Questions, models and tables that would result in invalid joins should not be shown in the entity picker for joins.

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:127.0) Gecko/20100101 Firefox/127.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.20.1+1",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.20.1",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.20.1+1",
    "os.name": "Mac OS X",
    "os.version": "14.2.1",
    "user.language": "en",
    "user.timezone": "Europe/Amsterdam"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.214 (2022-06-13)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.214 (2022-06-13)"
      }
    },
    "run-mode": "dev",
    "version": {
      "date": "2024-06-28",
      "src_hash": "af14ffe6d342f2c32114b0cac8d56b060c3bcd29",
      "tag": "v1.1.15-SNAPSHOT",
      "hash": "da0c9cb"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

P3

### Additional context

This issue is a follow up for https://github.com/metabase/metabase/issues/38989

### PR DESCRIPTION
**Pull request by escherize**
fixes https://github.com/metabase/metabase/issues/44974

- adds database_id to recents, and question collection items
- adds db id filtering to recents

sameDBJoins

### CODE REVIEW

File: e2e/test/scenarios/question-reproductions/reproductions-3.cy.spec.js [hunk 1]
```diff
@@ -1598,9 +1597,9 @@ describe.skip("issue 44974", () => {
...
     withDatabase(PG_DB_ID, ({ PEOPLE_ID }) => {
       const questionDetails = {
-        name: "Question 44794 in Postgres DB",
+        name: "Question 44974 in Postgres DB",
+        database: PG_DB_ID,
         query: {
-          database: PG_DB_ID,
           "source-table": PEOPLE_ID,
           limit: 1,
         },
```
Comment (iethree) [inline e2e/test/scenarios/question-reproductions/reproductions-3.cy.spec.js:1605]: @romeovs FYI, `database` goes outside the query 😁

**Review by noahmoss — APPROVED** [review]

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx [hunk 2]
```diff
@@ -278,6 +284,39 @@ describe("Notebook Editor > Join Step", () => {
     ).not.toBeInTheDocument();
   });
 
+  it("questions from another database should not appear in recents", async () => {
+    setup(undefined, {
+      readOnly: false,
+      recentItems: [
+        createMockRecentCollectionItem({
+          model: "card",
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx:292]: :x: It's a model. This change will make the test fail.



```suggestion
          model: "dataset",
```

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx [hunk 3]
```diff
@@ -278,6 +284,39 @@ describe("Notebook Editor > Join Step", () => {
...
+    });
+
+    await userEvent.click(
+      within(screen.getByLabelText("Right table")).getByRole("button"),
+    );
+    const modal = await screen.findByTestId("entity-picker-modal");
+
+    await waitForLoaderToBeRemoved();
+
+    expect(within(modal).getByText("Recents")).toBeInTheDocument();
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx:313]: Let's also assert that this tab is the one currently active.

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx [hunk 4]
```diff
@@ -278,6 +284,39 @@ describe("Notebook Editor > Join Step", () => {
     ).not.toBeInTheDocument();
   });
 
+  it("questions from another database should not appear in recents", async () => {
+    setup(undefined, {
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx:288]: Awkward `undefined`. Looks like it's time to refactor `setup` to only accept a single param object.

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx [hunk 5]
```diff
@@ -278,6 +284,39 @@ describe("Notebook Editor > Join Step", () => {
     ).not.toBeInTheDocument();
   });
 
+  it("questions from another database should not appear in recents", async () => {
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinStep.unit.spec.tsx:287]: Please always include the tag for repros :pray: 

```suggestion
  it("questions from another database should not appear in recents (metabase#44974)", async () => {
```

File: frontend/src/metabase-types/api/activity.ts [hunk 6]
```diff
@@ -43,6 +43,7 @@ export interface RecentTableItem extends BaseRecentItem {
 export interface RecentCollectionItem extends BaseRecentItem {
   model: "collection" | "dashboard" | "card" | "dataset" | "metric";
   can_write: boolean;
+  database_id?: number; // for models and questions
```
Comment (kamilmielnik) [inline frontend/src/metabase-types/api/activity.ts:46]:
```suggestion
  database_id?: DatabaseId; // for models and questions
```

**Review by kamilmielnik — APPROVED** [review]
:+1:

**Pull request merged**